applications/TopologyOptimizationApplication/custom_utilities/topology_updating_utilities.h

```diff
@@ -0,0 +1,342 @@
+
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Solution utility that updates response values for next iteration.
+/** Detail class definition.
+
+ */
+
+class TopologyUpdatingUtilities
+{
+public:
+
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of TopologyUpdatingUtilities
+    KRATOS_CLASS_POINTER_DEFINITION(TopologyUpdatingUtilities);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    TopologyUpdatingUtilities( ModelPart& model_part )
+    : mrModelPart(model_part)
+    {
+    }
+
+    /// Destructor.
+    virtual ~TopologyUpdatingUtilities()
+    {
+    }
+
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    // ---------------------------------------------------------------------------------------------------------------------------------------------
+    // --------------------------------- UPDATE DENSITIES  -----------------------------------------------------------------------------------------
+    // ---------------------------------------------------------------------------------------------------------------------------------------------
+
+    /// Finds the value of the X_PHYS (density) and updates it into the optimization problem
+    void UpdateDensitiesUsingOCMethod( char update_type[], double volfrac, double greyscale , double OptItr , double qmax)
+    {
+        KRATOS_TRY;
+
+        if ( strcmp( update_type , "oc_algorithm" ) == 0 ){
+            BuiltinTimer timer;
+            KRATOS_INFO("[TopOpt]") << "  Optimality Criterion Method (OC) chosen to solve the optimization problem" << std::endl;
+
+            // Check if Grey Scale Filter should be used
+            double q = 1;
+            if (greyscale == 1)
+            {
+                if (OptItr < 10)
+                    q = 1;
+                else
+                    q = std::min(qmax, 1.1*q);
+
+                KRATOS_INFO("[TopOpt]") << "  Grey Scale Filter activated, q = " << q << std::endl;
+            }
+            else
+            {
+                KRATOS_INFO("[TopOpt]") << "  Grey Scale Filter deactivated, q = " << q << std::endl;
+            }
+
+            // Update Densities procedure
+            double l1    = 0.0;
+            double l2    = 1.0e12; ///10000000000.0
+            double move    = 0.2;
+            double sum_X_Phys;
+            int nele;
+            double x_new = 0.0;
+            double lmid = 0.0;
+
+            // Bisection algorithm to find Lagrange Multiplier so that volume constraint is satisfied (lmid)
+            while ((l2-l1)/(l1+l2) > 0.001)
+            {
+                lmid = 0.5*(l2+l1);
+                sum_X_Phys = 0.0;
+                nele = 0;
+                x_new = 0.0;
+
+                for( ModelPart::ElementIterator element_i = mrModelPart.ElementsBegin(); element_i!= mrModelPart.ElementsEnd(); element_i++ )
```

Comment:
Parallel?

Reply:
Idem

Reply:
In this case and in the case of structure_response_function.h I think it is better to not use parallel utilities. When I use parallel utilities here I get bad results. Due to the fact that the elements are updated and summed up, something is going wrong if it is not done iteratively.